src/data/real/ereal.lean

```diff
@@ -0,0 +1,159 @@
+/-
+Copyright (c) 2019 Kevin Buzzard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin Buzzard
+-/
+
+import data.real.basic 
+
+/-!
+# The extended reals [-∞, ∞].
+
+This file defines `ereal`, the real numbers together with a top and bottom element,
+referred to as ⊤ and ⊥. 
+
+Addition and multiplication are problematic in the presence of ±∞, but
+negation is not, so we define negation. The main work is defining
+Sup : set ereal → ereal and proving that it is a Sup in all cases.
+Once we have, this, it is not hard to give a complete lattice
+structure on `ereal`.
+
+## Tags
+
+real, ereal, complete lattice
+-/
+
+/-- ereal : The type $$[-\infty,+\infty]$$ or `[-∞, ∞]` -/
+def ereal := with_bot (with_top ℝ)
+
+instance : linear_order ereal := by unfold ereal; apply_instance
+instance : lattice.order_bot ereal := by unfold ereal; apply_instance
+instance : lattice.order_top ereal := by unfold ereal; apply_instance
+
+/- neg -/
+
+/-- negation on ereal -/
+def ereal.neg : ereal → ereal
+| none := ⊤
+| (some none) := ⊥
+| (some (some x)) := ((↑-x : with_top ℝ) : with_bot (with_top ℝ))
+
+instance : has_neg ereal := ⟨ereal.neg⟩
+
+/-- if -a ≤ b then -b ≤ a on ereal -/
+theorem ereal.neg_le_of_neg_le : ∀ {a b : ereal} (h : -a ≤ b), -b ≤ a
+| none none h := by cases (lattice.le_bot_iff.1 h)
+| none (some b) h := by cases (lattice.top_le_iff.1 h); exact le_refl _
+| (some none) b h := lattice.le_top
+| (some (some a)) none h := by cases (lattice.le_bot_iff.1 h)
+| (some (some a)) (some none) h := lattice.bot_le
+| (some (some a)) (some (some b)) h := 
+begin
+  revert h,
+  change (((-a : ℝ) : with_top ℝ) : with_bot (with_top ℝ)) ≤ _ →
+    (((-b : ℝ) : with_top ℝ) : with_bot (with_top ℝ)) ≤ _,
+  unfold_coes, simpa using neg_le_of_neg_le,
+end
+
+/-- -a ≤ b ↔ -b ≤ a on ereal-/
+theorem ereal.neg_le {a b : ereal} : -a ≤ b ↔ -b ≤ a := ⟨ereal.neg_le_of_neg_le, ereal.neg_le_of_neg_le⟩
+
+/-- - -a = a on ereal -/
+theorem ereal.neg_neg : ∀ (a : ereal), - (- a) = a
+| none := rfl
+| (some none) := rfl
+| (some (some a)) := show (((- -a : ℝ) : with_top ℝ) : with_bot (with_top ℝ)) = (((a : ℝ) : with_top ℝ) : with_bot (with_top ℝ)),
+by simp [neg_neg a]
+
+/-- a ≤ -b → b ≤ -a on ereal -/
+theorem ereal.le_neg_of_le_neg {a b : ereal} (h : a ≤ -b) : b ≤ -a :=
+by rwa [←ereal.neg_neg b, ereal.neg_le, ereal.neg_neg]
+
+/-- The claim that a set of ereals has a supremum in ereal -/
+def has_Sup (X : set ereal) : Prop := ∃ l : ereal, is_lub X l
+
+local attribute [instance, priority 10] classical.prop_decidable
+
+/-- A set of ereals has a Sup in ereal -/
+theorem Sup_exists (X : set ereal) : has_Sup X :=
+  let Xoc : set (with_top ℝ) := λ x, X (↑x : with_bot _) in
+dite (Xoc = ∅) (λ h, ⟨⊥, ⟨
```

Comment:
Is there a reason to use `dite` instead of its sugared version?

Reply:
I needed a dependent if/then/else. What's the sugared way to do this?

Reply:
`if h : Xoc = ∅ then _ else _`

Reply:
Done in ac2ac59bac50dce950a345104b09494a7daee19f